File: 01-50/14/14.c
#include <stdio.h>
#include <stdint.h>

int64_t chain_length(start) {
    int64_t len = 1;
    int64_t n = start;
    while (n > 1) {
        len++;
        if (n % 2 == 0) {
            n = n / 2;
        } else {
            n = (3 * n) + 1;
        }
    }
    return len;
}

int main() {
    struct {
        int64_t n, len;
    } best = {0};
    for (int64_t i = 0; i < 1000000; i++) {
        int64_t len = chain_length(i);
        if (len >= best.len) {
            best.n = i;
            best.len = len;
        }
    }
    printf("%lld = %lld\n", best.n, best.len);
}

File: 01-50/14/14.py
def length1(start):
    length = 1
    n = start
    while n > 1:
        length += 1
        if n % 2 == 0:
            n = n / 2
        else:
            n = 3 * n + 1
    return length


def length2(start):
    length = 1
    n = start
    while n > 1:
        while n & 1:
            n = (3 * n + 1) / 2
            length += 2
        while not n & 1:
            n /= 2
            length += 1
    return length


best_len = 0
for i in xrange(0, 1000000):
    print i
    length = length1(i)
    if length > best_len:
        best = i
        best_len = length

print best, best_len
